zipkin-thrift/src/main/thrift/com/twitter/zipkin/zipkinCore.thrift

```diff
@@ -15,18 +15,118 @@ namespace java com.twitter.zipkin.thriftjava
+ */
 const string CLIENT_RECV = "cr"
+/*
+ * The server sent ("ss") a response to a client. There is only one response
+ * per span. If there's a transport error, each attempt can be logged as a
+ * WIRE_SEND annotation.
+ *
+ * Typically, a trace ends with a server send, so the last timestamp of a trace
+ * is often the timestamp of the root span's server send.
+ *
+ * If chunking is involved, each chunk could be logged as a separate
+ * SERVER_SEND_FRAGMENT in the same span.
+ *
+ * Annotation.host is not the client. It is the host which logged the send
+ * event, almost always the server.
+ *
+ * To record the actual endpoints of the client and server side of the span,
+ * use CLIENT_ADDR and SERVER_ADDR respectively.
+ */
 const string SERVER_SEND = "ss"
+/*
+ * The server received ("cr") a request from a client. There is only one
+ * request per span.  For example, if duplicate responses were received, each
+ * can be logged as a WIRE_RECV annotation.
+ *
+ * Typically, a trace starts with a server receive, so the first timestamp of a
+ * trace is often the timestamp of the root span's server receive.
+ *
+ * If chunking is involved, each chunk could be logged as a separate
+ * SERVER_RECV_FRAGMENT in the same span.
+ *
+ * Annotation.host is not the client. It is the host which logged the send
+ * event, almost always the server.
+ *
+ * To record the actual endpoints of the client and server side of the span,
+ * use CLIENT_ADDR and SERVER_ADDR respectively.
+ */
 const string SERVER_RECV = "sr"
+/*
+ * Optionally logs an attempt to send a message on the wire. Multiple wire send
+ * events could indicate network retries. A lag between client or server send
+ * and wire send might indicate queing or processing delay.
+ */
 const string WIRE_SEND = "ws"
+/*
+ * Optionally logs an attempt to receive a message from the wire. Multiple wire
+ * receive events could indicate network retries. A lag between wire receive
+ * and client or server receive might indicate queing or processing delay.
+ */
 const string WIRE_RECV = "wr"
+/*
+ * Optionally logs progress of a (CLIENT_SEND, WIRE_SEND). For example, this
+ * could be one chunk in a chunked request.
+ */
+const string CLIENT_SEND_FRAGMENT = "csf"
+/*
+ * Optionally logs progress of a (CLIENT_RECV, WIRE_RECV). For example, this
+ * could be one chunk in a chunked response.
+ */
+const string CLIENT_RECV_FRAGMENT = "crf"
+/*
+ * Optionally logs progress of a (SERVER_SEND, WIRE_SEND). For example, this
+ * could be one chunk in a chunked response.
+ */
+const string SERVER_SEND_FRAGMENT = "ssf"
+/*
+ * Optionally logs progress of a (SERVER_RECV, WIRE_RECV). For example, this
+ * could be one chunk in a chunked request.
+ */
+const string SERVER_RECV_FRAGMENT = "srf"
 
-#************** Common binary annotation keys **************
-// The endpoint associated with CLIENT_ annotations is not necessarily Annotation.host
+#***** BinaryAnnotation.key where value = [1] and annotation_type = BOOL ******
```

Comment:
First, if it isn't clear, I am documenting what actually exists, for at
least 2 years. Sadly, a lot of instrumentation including Ruby, wasn't well
maintained before you arrived. It might be based on 3 years ago. Let me try
to explain, and please keep an open mind. We have a separate work to making
a new zipkin model. This is about making existing ones compatible.

Zipkin is a logging system and annotations are log lines in that. The host
of an annotation is the host that logged, in this case a client send.
Because these events are placed in time, I can use the host dimension to
find out more about what hosts were doing, such as making client requests.
Knowing this, it is very surprising to put a host in an annotation that
wasn't the actual host. For example, I cannot group my
log/annotation/events by ip anymore.

Binary annotations are tags on a span. It is where you add extra metadata.
For example, you may have multiple ways of establishing the client address,
for example the x-forwarded-for header. Whatever you can find out, can be
added here, and would apply to 'the entire span'. For example, I could make
a query for all events dealing with a client by looking for a single
key/value pair (CA). That might include several actual hosts!

How a client or server address is encoded is confusing. I agree. Because it
is, feel free to read below only if you really care ;) Otherwise, I hope we
can move past the past. If we want to make zipkin's model better, we need
to at least understand what it does now. The Ruby and also other code have
done incorrect things over time because there was no documentation. This
isnt changing anything except that it is now documented. Make sense?

-- why single byte for CA/SA value

For example, it takes a while to understand why bool (a byte array of size
one) was chosen back in 2013. The reason is that there's already a struct
to hold the endpoint and it is already a member of binary annotation. Since
binary annotation.value is not optional, what's the smallest thing you can
place here? Well.. A single byte is the answer to that.
On Oct 25, 2015 12:01 AM, "Jordi Polo Carres" <notifications@github.com>
wrote:

> In zipkin-thrift/src/main/thrift/com/twitter/zipkin/zipkinCore.thrift
> <https://github.com/openzipkin/zipkin/pull/794#discussion_r42941888>:
>
> >
> > -#************** Common binary annotation keys **************
> > -// The endpoint associated with CLIENT_ annotations is not necessarily Annotation.host
> > +#***** BinaryAnnotation.key where value = [1] and annotation_type = BOOL ******
>
> Agreed. What's the meaning of having endpoint sent all the time if there
> is no use for it and the canonical source is ca sa ?
> In fact makes it confusing.
>
> The original ruby code was assuming cr cs should have the servicename and
> addresses of the server which was called. Kind of makes sense. And the UI
> looks good following that assumption also.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openzipkin/zipkin/pull/794/files#r42941888>.
>